plonky2x/src/frontend/vars/array.rs

```diff
@@ -162,18 +168,92 @@ impl<L: PlonkParameters<D>, const D: usize> CircuitBuilder<L, D> {
         V::from_variables_unsafe(&selected_vars)
     }
 
-    pub fn get_fixed_subarray<const SIZE: usize>(
+    pub fn get_fixed_subarray<const ARRAY_SIZE: usize, const SUB_ARRAY_SIZE: usize>(
         &mut self,
-        array: &[Variable],
-        start: Variable,
-    ) -> ArrayVariable<Variable, SIZE> {
-        let mut res = Vec::new();
-        for i in 0..SIZE {
-            let i_variable = self.constant(L::Field::from_canonical_usize(i));
-            let index = self.add(start, i_variable);
-            res.push(self.select_array(array, index));
+        array: &ArrayVariable<Variable, ARRAY_SIZE>,
+        start_idx: Variable,
+        seed: &[ByteVariable],
+    ) -> ArrayVariable<Variable, SUB_ARRAY_SIZE> {
+        const MIN_SEED_BITS: usize = 120; // TODO: Seed it with 120 bits.  Need to figure out if this is enough bits of security.
+
+        let mut input_stream = VariableStream::new();
+        for array_element in array.as_vec().iter() {
+            input_stream.write(array_element);
+        }
+        input_stream.write(&start_idx);
+
+        let hint = SubArrayExtractorHint {
+            array_size: ARRAY_SIZE,
+            sub_array_size: SUB_ARRAY_SIZE,
+        };
+        let output_stream = self.hint(input_stream, hint);
+
+        let mut sub_array_element = Vec::new();
+        for _i in 0..SUB_ARRAY_SIZE {
+            sub_array_element.push(output_stream.read::<Variable>(self));
+        }
+
+        let sub_array = ArrayVariable::<Variable, SUB_ARRAY_SIZE>::from(sub_array_element);
+
+        let mut seed_targets = Vec::new();
+        let mut challenger = RecursiveChallenger::<L::Field, PoseidonHash, D>::new(&mut self.api);
+
+        // Need to get chunks of 7 since the max value of F is slightly less then 64 bits.
+        let mut seed_bit_len = 0;
+        for seed_chunk in seed.to_vec().chunks(7) {
+            let seed_element_bits = seed_chunk
+                .iter()
+                .flat_map(|x| x.as_bool_targets())
+                .collect_vec();
+            let seed_element = self.api.le_sum(seed_element_bits.iter());
+            seed_bit_len += seed_element_bits.len();
+            seed_targets.push(seed_element);
         }
-        ArrayVariable::new(res)
+
+        assert!(seed_bit_len >= MIN_SEED_BITS);
+
+        challenger.observe_elements(seed_targets.as_slice());
```

Comment:
Why can't we just do `challenger.observe_elements(seed.to_vec().flat_map(|x| x.as_targets()).collect()` there seems to be no need to convert it into elements and add additional constraints because the total amount of randomness will be the same whether it's in bit form or byte form?

Reply:
Ya, I agree that's cleaner.

I think packing the seed into golidlocks field could potentially reduce the total overall number of constraints.  An iteration of poseidon does get kicked off after a set number of elements are observed.